### ISSUE TITLE
Using STL extensions with Clang on Linux

### ISSUE THREAD

**Issue by tdelame** [PixarAnimationStudios/USD#1291]
### Description of Issue
**TL;DR:** provide a way to turn ARCH_HAS_GNU_STL_EXTENSIONS on when using Clang on Linux. 

**Long Version:** In the project I am working on, we create all our binaries with gcc on Linux. However, to have access to more development tools, have less warnings (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=53431) and more expressive error messages, we use clang with libstdc++ during the development. Our Usd package is thus build with gcc on Linux, which means ARCH_HAS_GNU_STL_EXTENSIONS is turned on. So in our Usd binaries, TfHashMap and TfHashSet use STL extensions and both have a size of 40. When compiling with Clang, ARCH_HAS_GNU_STL_EXTENSIONS is not defined, and TfHashMap and TfHashSet both have a size of 56, which leads to crash when linking with our Usd package. If we force ARCH_HAS_GNU_STL_EXTENSIONS to be defined (by editing `include/pxr/base/arch/defines.h`), everything works fine.

Note: `<ext/hash_map>` is tagged as deprecated in gcc 10.1.0.

### Steps to Reproduce
1. Edit `include/pxr/base/arch/defines.h` to replace the line
```
#if defined(ARCH_OS_LINUX) && defined(ARCH_COMPILER_GCC)
```
with 
```
#if defined(ARCH_OS_LINUX) && (defined(ARCH_COMPILER_GCC) || defined(USD_FORCE_GNU_STL_EXTENSIONS))
```
2. Write a file reproduction.cpp with the following content
```
# include <iostream>
# include <pxr/base/tf/hashmap.h>
# include <pxr/base/tf/hashset.h>

int main() {
  std::cout << "sizeof(TfHashSet) = " << sizeof(pxr::TfHashSet<int>) << '\n'
            << "sizeof(TfHashMap) = " << sizeof(pxr::TfHashMap<int, int>) << std::endl;
  return 0;
}
```
3. Compile the reproduction file with `g++ -D_GLIBCXX_USE_CXX11_ABI=0 -std=c++14 -isystem ${USD_INSTALLATION}/include  reproduction.cpp -o reproduction` and run it. The displayed sizes are both 40 octets.
4. Compile the reproduction file with `clang++ -D_GLIBCXX_USE_CXX11_ABI=0 -std=c++14 -stdlib=libstdc++ -isystem ${USD_INSTALLATION}/include reproduction.cpp -o reproduction` and run it. The displayed sizes are both 56 octets.
5. Compile the reproduction file with `clang++ -D_GLIBCXX_USE_CXX11_ABI=0 -std=c++14 -stdlib=libstdc++ -DUSD_FORCE_GNU_STL_EXTENSIONS -isystem ${USD_INSTALLATION}/include reproduction.cpp -o reproduction` and run it. The displayed sizes are both 40 octets.

### System Information (OS, Hardware)
Linux.

### Package Versions
Tested on 20.05.

**Comment by jtran56** [comment]
Filed as internal issue #USD-6285

**Comment by jwnimmer-tri** [comment]
Given that `<ext/hash_map>` is long since deprecated, instead of doubling down on it and allowing Clang opt-in, I'd recommend getting rid of`ARCH_HAS_GNU_STL_EXTENSIONS` entirely and therefore using `<unordered_map>` and `<unordered_set>` uniformly on all platforms.
